README.md
# connect-mongostore

  MongoDB session store for Connect/Express

  [![Build Status](https://secure.travis-ci.org/diversario/connect-mongostore.png?branch=master)](http://travis-ci.org/diversario/connect-mongostore)
  [![Coverage Status](https://coveralls.io/repos/diversario/connect-mongostore/badge.png?branch=master)](https://coveralls.io/r/diversario/connect-mongostore?branch=master)
  [![Dependency Status](https://gemnasium.com/diversario/connect-mongostore.png)](https://gemnasium.com/diversario/connect-mongostore)
  [![NPM version](https://badge.fury.io/js/connect-mongostore.png)](http://badge.fury.io/js/connect-mongostore)
  
### Why?
  
  Existing stores work fine but we needed a store that supports replica sets configuration passed to it.

## Installation

connect-mongostore supports connect `>= 1.0.3`, express `3.x` and express `4.x` with express-session.

via npm:

    $ npm install connect-mongostore

## Options

  Pass a fully-qualified URI as the only option (e.g., `mongodb://127.0.0.1:27017/myDatabase`), or an object (also see `examples` folder for configuration examples):

  - `db` Can be three different things:
    - database name (string)
    - mongo-native database instance
    - object with replica set options. These options requires:
    
      + `name` Database name
      + `servers` Array of replica set server configurations similar to:

          ```javascript
            {
              "host" : "127.0.0.1", // required
              "port" : 27017, // required
              "options" : { // all optional
                "autoReconnect" : false,
                "poolSize" : 200,
                "socketOptions" : {
                  "timeout" : 0,
                  "noDelay" : true,
                  "keepAlive" : 1,
                  "encoding" : "utf8"
                }
              }
            }
          ```
          Configuration options explained [here](http://mongodb.github.io/node-mongodb-native/markdown-docs/database.html)
      + `replicaSetOptions` An object with a single `rs_name` property specifying your replica set name
  - `collection` Collection (optional, default: `sessions`) 
  - `host` MongoDB server hostname (optional, default: `127.0.0.1`). Not needed for Replica Sets.
  - `port` MongoDB server port (optional, default: `27017`) Not needed for Replica Sets.
  - `username` Username (optional)
  - `password` Password (optional)
  - `expireAfter` Duration of session cookies in milliseconds (e.g., ones with `maxAge` not defined). Defaults to 2 weeks.
    May be useful if you see a lot of orphaned sessions in the database and want them removed sooner than 2 weeks.
  - `autoReconnect` This is passed directly to the MongoDB `Server` constructor as the auto_reconnect
                     option (optional, default: false).
  - `ssl` Use SSL to connect to MongoDB (optional, default: false).
  - `mongooseConnection` in the form: `mongooseDatabase.connections[0]` to use an existing mongoose connection. (optional)
  - `stringify` If false, connect-mongostore will serialize sessions using `JSON.stringify` before
                setting them, and deserialize them with `JSON.parse` when getting them.
                (optional, default: false). Note that deserialization will not revive Dates, Object IDs and other non-plain objects.

The second parameter to the `MongoStore` constructor is a callback which will be called once the database connection is established.
This is mainly used for the tests, however you can use this callback if you want to wait until the store has connected before
starting your app.

## Example

With express 3.x:

    var express = require('express');
    var MongoStore = require('connect-mongostore')(express);
    var app = express();
    
    app.use(express.session({
        secret: 'my secret',
        store: new MongoStore({'db': 'sessions'})
      }));

With express 4.x:

    var express = require('express');
    var session = require('express-session');
    var MongoStore = require('connect-mongostore')(session);
    var app = express();

    app.use(session({
        secret: 'my secret',
        store: new MongoStore({'db': 'sessions'})
      }));

With connect:

    var connect = require('connect');
    var MongoStore = require('connect-mongostore')(connect);

## Removing expired sessions

  connect-mongostore uses MongoDB's TTL collection feature (2.2+) to
  have mongod automatically remove expired sessions. (mongod runs this
  check every minute.)

  **Note:** By connect/express's default, session cookies are set to 
  expire when the user closes their browser (maxAge: null). In accordance
  with standard industry practices, connect-mongostore will set these sessions
  to expire two weeks from their last 'set'. You can override this 
  behavior by manually setting the maxAge for your cookies - just keep in
  mind that any value less than 60 seconds is pointless, as mongod will
  only delete expired documents in a TTL collection every minute.

  For more information, consult connect's [session documentation](http://www.senchalabs.org/connect/session.html).

## Tests

You need `mocha`.

    make test

The tests use a database called `connect-mongostore-test`. `make test` **does not** run replica set tests.

To run all tests including replica set tests:

    make test-rs

Note that replica set tests will fail unless you 1) have a replica set, and 2) set the address of that replica set either in `CM_REPL_SET_HOST` environment variable or directly in `connect-mongostore.test.js` file.

You can check code coverage report by running

    make coverage
    
or
    
    make coverage-rs
        
for coverage with replica set tests.
    
Coverage report will be in `reports/lcov-report/index.html` file.


## Stuff

Big thanks to @kcbanner and his `connect-mongo`, which was a starting point for `connect-mongostore`.

## License 

(The MIT License)


Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
0.1.0 / 2014-04-14 
==================

 * Add support for Express 4.x by passing in reference to `session` instead of reference to `express`
 * Updated Readme to reflect the updates and support for Express
 * Add History.md to track changes


test/connect-mongostore.test.js
'use strict'

var assert = require('assert')
  , crypto = require('crypto')

var connect = require('connect')
  , mongoose = require('mongoose')
  , mongo = require('mongodb')
  , _ = require('lodash')

var MongoStore = require('../.')(connect)

var defaultOptions = {'w': 1, 'host': '127.0.0.1', 'port': 27017, 'autoReconnect': true, 'ssl': false}
  , dbName = 'connect-mongostore-test' 
  , collectionName = 'sessions'
  , defaultDbOptions = {'db': dbName}
  , expirationPeriod = 1000 * 60 * 60 * 24 * 14 // 2 weeks

var mongooseDb = {'mongooseConnection': mongoose.connect('mongodb://127.0.0.1:27017/' + dbName).connections[0]}
  , mongoNativeDb = {'db': new mongo.Db(dbName, new mongo.Server('127.0.0.1', 27017, {}), {'w': defaultOptions.w})}

var replSetMember = {
  "host" : process.env.CM_REPL_SET_HOST || "192.168.1.225",
  "port" : -1,
  "options" : {
    "autoReconnect" : false,
    "poolSize" : 200,
    "socketOptions" : {
      "timeout" : 0,
      "noDelay" : true,
      "keepAlive" : 1,
      "encoding" : "utf8"
    }
  }
}

var replSetConfig = {
  "collection" : collectionName,
  "db": {
    "name" : dbName,
    "servers" : [
      getReplSetMember(27017),
      getReplSetMember(27018),
      getReplSetMember(27019)
    ],
    'replicaSetOptions' : {
      'rs_name' : 'rs0',
      'w': 'majority',
      'read_secondary': true
    }
  }
}


function getReplSetMember(port) {
  var config = JSON.parse(JSON.stringify(replSetMember))
  config.port = port
  return config
}


function dbFromMongooseConnection(mongooseConnection, opts) {
  if (mongooseConnection.user && mongooseConnection.pass) {
    opts.username = mongooseConnection.user
    opts.password = mongooseConnection.pass
  }

  return new mongo.Db(
    mongooseConnection.db.databaseName,
    new mongo.Server(mongooseConnection.db.serverConfig.host,
      mongooseConnection.db.serverConfig.port,
      mongooseConnection.db.serverConfig.options),
    { w: opts.w || defaultOptions.w }
  )
}


/**
 * Performs authentication, if credentials provided
 * @param db
 * @param options
 * @param cb
 */
function authenticate(db, options, cb) {
  if (options.username && options.password) {
    db.admin().authenticate(options.username, options.password, cb)
  } else cb()
}



/**
 * Returns a collection reference
 * @param db
 * @param cb
 */
function getCollection(db, cb) {
  db.collection(collectionName, function (err, collection) {
    cb(err, collection)
  })
}



/**
 * Creates and opens a database instance
 * @param opts
 * @param cb
 * @returns {*}
 */
function openDatabase(opts, cb) {
  var db

  if (opts.mongooseConnection) {
    db = dbFromMongooseConnection(opts.mongooseConnection, opts)
  } else if (!opts.db) {
    throw new Error('Required MongoStore option `db` missing')
  } else if (typeof opts.db == 'object' && opts.db.databaseName) { // Assume it's an instantiated DB Object
    db = opts.db
  } else if (Array.isArray(opts.db.servers)) {
    var serverArray = [];

    opts.db.servers.forEach(function (server) {
      var serverOptions = server.options || {}
      var newServer = new mongo.Server(server.host, server.port, serverOptions)
      serverArray.push(newServer)
    })

    db = new mongo.Db(opts.db.name, new mongo.ReplSetServers(serverArray), opts.db.replicaSetOptions || {w: 0})
  } else if (typeof opts.db != 'string') {
    throw new Error('`db` option must be a string, array or a database instance.')
  } else {
    db = new mongo.Db(
      opts.db, new mongo.Server(
        opts.host || defaultOptions.host,
        opts.port || defaultOptions.port,
        {
          'auto_reconnect': opts.autoReconnect || defaultOptions.autoReconnect,
          'ssl': opts.ssl || defaultOptions.ssl
        }),
      {'w': opts.w || defaultOptions.w})
  }

  if (db.openCalled) return cb(db)
  
  db.open(function (err) {
    cb(db)
  })
}



/**
 * Returns instances of store, mongo-native database and
 * mongo-native collection.
 * @param options
 * @param cb
 */
function getInstances(options, cb) {
  var store = new MongoStore(options)
  openDatabase(options, function (db) {
    authenticate(db, options, function () {
      getCollection(db, function (err, coll) {
        cb(store, db, coll)
      })
    })
  })
}


function closeStore(store) {
  store.db.close()
}

function cleanup(store, db, collection, cb) {
  collection.drop(function (err, result) {
    db.close()
    closeStore(store)
    cb && cb()
  })
}


function getRandomString() {
  return crypto.randomBytes(16).toString('hex')
}


describe('Connect-mongostore', function () {
  function tests(ctorOptions, suiteCallback) {
    describe('Basic stuff:', function () {
      var store, db, collection

      function init(opts, done) {
        if (!done) {
          done = opts
          opts = [{}]
        }

        if (!Array.isArray(opts)) opts = [opts]

        opts.push(ctorOptions)

        var config = {}

        opts.forEach(function (opt) {
          _.extend(config, opt)
        })


        getInstances(config, function (_store, _db, _collection) {
          store = _store
          db = _db
          collection = _collection
          done()
        })
      }

      afterEach(function (done) {
        cleanup(store, db, collection, function () {
          done()
        })
      })
      
      it('after first call getCollection returns the cached collection', function (done) {
        init(function () {
          store.getCollection(function (coll) {
            assert(coll)
            
            coll.__JUST_TESTING = true
            
            store.getCollection(function (coll) {
              assert(coll)
              assert(coll.__JUST_TESTING)
              done()
            })            
          })
        })
      })
    })
    
    describe('Set', function () {
      var store, db, collection
    
      function init(opts, done) {
        if (!done) {
          done = opts
          opts = [{}]
        }
    
        if (!Array.isArray(opts)) opts = [opts]
    
        opts.push(ctorOptions)
    
        var config = {}
    
        opts.forEach(function (opt) {
          _.extend(config, opt)
        })
    
        
        getInstances(config, function (_store, _db, _collection) {
          store = _store
          db = _db
          collection = _collection
          done()
        })
      }
    
      afterEach(function (done) {
        cleanup(store, db, collection, function () {
          done()
        })
      })
      
      it('saves session as object', function (done) {
        init(function () {
          var sid = getRandomString()
            , s = {'test': getRandomString()}
            , date = new Date
          
          store.set(sid, s, function (err, session) {
            assert.strictEqual(err, null)
    
            collection.findOne({'_id': sid}, function (err, session) {
              assert(session.expires - date < expirationPeriod + 100)
              delete session.expires
              
              assert.deepEqual(session, {
                'session': s,
                '_id': sid
              })
              
    
              done()
            })
          })
        })
      })
    
      it('saves session as string', function (done) {
        init({'stringify': true}, function () {
          var sid = getRandomString()
            , s = {'test': getRandomString()}
            , date = new Date
          
          store.set(sid, s, function (err, session) {
            assert(!err)
      
            collection.findOne({'_id': sid}, function (err, session) {
              assert(session.expires - date < expirationPeriod + 100)
              delete session.expires
              
              assert.deepEqual(session, {
                'session': JSON.stringify(s),
                '_id': sid
              })
    
              done()
            })
          })
        })
      })
      
      it('saves session with expiration', function (done) {
        init(function () {
          var sid = getRandomString()
          var s = {
            'test': getRandomString(),
            'cookie': {
              '_expires': new Date
            }
          }
    
          store.set(sid, s, function (err, session) {
            assert.strictEqual(err, null)
    
            collection.findOne({'_id': sid}, function (err, session) {
              assert.deepEqual(session.session, s)
              assert.strictEqual(session._id, sid)
              assert.equal(session.expires.toJSON(), new Date(s.cookie._expires).toJSON())
    
              done()
            })
          })
        })
      })
      
      it('saves session with default expiration', function (done) {
        init(function () {
          var sid = getRandomString()
          var s = {
            'test': getRandomString(),
            'cookie': {}
          }

          store.set(sid, s, function (err, session) {
            assert.strictEqual(err, null)
            
            var expirationDate = new Date(Date.now() + 1000 * 60 * 60 * 24 * 14) // two weeks

            collection.findOne({'_id': sid}, function (err, session) {
              assert.deepEqual(session.session, s)
              assert.strictEqual(session._id, sid)
              
              assert(expirationDate - session.expires < 1000) // a generous 1 second for stuff to run

              done()
            })
          })
        })
      })

      it('saves session with overriden default expiration', function (done) {
        init({'expireAfter': 10000}, function () {
          var sid = getRandomString()
          var s = {
            'test': getRandomString(),
            'cookie': {}
          }

          store.set(sid, s, function (err, session) {
            assert.strictEqual(err, null)

            var expirationDate = new Date(Date.now() + 10000) // two weeks

            collection.findOne({'_id': sid}, function (err, session) {
              assert.deepEqual(session.session, s)
              assert.strictEqual(session._id, sid)

              assert(expirationDate - session.expires < 1000) // a generous 1 second for stuff to run

              done()
            })
          })
        })
      })
    })
    
    
    
    
    describe('Get', function () {
      var store, db, collection
    
      function init(opts, done) {
        if (!done) {
          done = opts
          opts = [{}]
        }
    
        if (!Array.isArray(opts)) opts = [opts]
    
        opts.push(ctorOptions)
        
        var config = {}
    
        opts.forEach(function (opt) {
          _.extend(config, opt)
        })
    
        getInstances(config, function (_store, _db, _collection) {
          store = _store
          db = _db
          collection = _collection
          done()
        })
      }
    
      afterEach(function (done) {
        cleanup(store, db, collection, function () {
          done()
        })
      })
      
      it('gets session', function (done) {
        init(function () {
          var sid = getRandomString()
            , s = {'one': getRandomString(), 'two': {'three': getRandomString()}, 'four': 42}
          
          collection.insert({'_id': sid, 'session': s}, function (error, ids) {
            store.get(sid, function (err, session) {
              assert.deepEqual(session, s)
              done()
            })
          })
        })
      })

      it('gets non-existent session', function (done) {
        init(function () {
          var sid = getRandomString()

          store.get(sid, function (err, session) {
            assert(!err)
            assert(!session)
            done()
          })
        })
      })
      
      it('gets session as string', function (done) {
        init({stringify: true}, function () {
          var sid = getRandomString()
            , s = {'one': getRandomString(), 'two': {'three': getRandomString()}, 'four': 42}
          
          collection.insert({'_id': sid, 'session': JSON.stringify(s)}, function (error, ids) {
            store.get(sid, function (err, session) {
              assert.deepEqual(session, s)
              done()
            })
          })
        })
      })
      
      it('gets length', function (done) {
        init(function () {
          var sid = getRandomString()
            , s = {'one': getRandomString(), 'two': {'three': getRandomString()}, 'four': 42}
          
          collection.insert({'_id': sid, 'session': s}, function (error, ids) {
            store.length(function (err, length) {
              assert(!err)
              assert.strictEqual(length, 1)
              done()
            })
          })
        })
      })
      
      it('destroys a session', function (done) {
        init(function () {
          var sid = getRandomString()
            , s = {'one': getRandomString(), 'two': {'three': getRandomString()}, 'four': 42}
          
          collection.insert({'_id': sid, 'session': s}, function (error, ids) {
            store.destroy(sid, function (err) {
              assert(!err)
              done()
            })
          })
        })
      })
    
      it('clears all sessions', function (done) {
        init(function () {
          var sid = getRandomString()
            , s = {'one': getRandomString(), 'two': {'three': getRandomString()}, 'four': 42}
          
          collection.insert({'_id': sid, 'session': s}, function (error, ids) {
            store.clear(function (err) {
              collection.count(function (err, count) {
                assert.strictEqual(count, 0)
                done()
              })
            })
          })
        })
      })

      it('does not return expired session', function (done) {
        init(function () {
          var sid = getRandomString()
            , s = {'one': getRandomString(), 'two': {'three': getRandomString()}, 'four': 42}

          collection.insert({'_id': sid, 'session': s, 'expires': new Date(Date.now()-1000)}, function (err, saved) {
            assert.deepEqual(saved[0].session, s)
            store.get(sid, function (err, session) {
              assert(!session)
              done()
            })
          })
        })
      })
    })
    
    
    
    describe('Options', function () {
      after(suiteCallback)
      
      it('support string URL', function (done) {
        var store = new MongoStore('mongodb://127.0.0.1:27017/connect-mongostore-test/sessions')
        assert.strictEqual(store.db.databaseName, dbName)
        assert.strictEqual(store.db.serverConfig.host, '127.0.0.1')
        assert.equal(store.db.serverConfig.port, 27017)
    
        store.getCollection(function () {
          assert.equal(store.collection.collectionName, collectionName)
          closeStore(store)
          done()
        })
      })
    
      it('support string URL with auth', function (done) {
        var store = new MongoStore('mongodb://test:test@127.0.0.1:27017/connect-mongostore-test/sessions')
    
        assert.strictEqual(store.db.databaseName, dbName)
        assert.strictEqual(store.db.serverConfig.host, '127.0.0.1')
        assert.equal(store.db.serverConfig.port, 27017)
    
        store.getCollection(function () {
          assert.equal(store.collection.collectionName, collectionName)
          closeStore(store)
          done()
        })
      })
    
      it('require some sort of database option', function () {
        assert.throws(function () {
          new MongoStore({})
        }, /`db` missing/)

        assert.throws(function () {
          new MongoStore({'mongooseConnection': 'foobar'}, function () {})
        }, Error)
        
        assert.throws(function () {
          var store = new MongoStore({'db': {}})
          store.getCollection()
        }, /`db` option must be/)
      })
      
      it('support options object with URL', function (done) {
        var store = new MongoStore({
          'url': 'mongodb://test:test@127.0.0.1:27017/',
          'db': dbName,
          'collection': collectionName
        })
    
        assert.strictEqual(store.db.databaseName, dbName)
        assert.strictEqual(store.db.serverConfig.host, '127.0.0.1')
        assert.equal(store.db.serverConfig.port, 27017)
    
        store.getCollection(function () {
          assert.equal(store.collection.collectionName, collectionName)
          closeStore(store)
          done()
        })
      })
    })
  }

  tests(defaultDbOptions, function () {
    tests(mongoNativeDb, function () {
      tests(mongooseDb, function () {
        if (process.env.HAS_JOSH_K_SEAL_OF_APPROVAL != 'true') { // do not run replica set tests on Travis CI
          tests(replSetConfig, function () {
          })
        }
      })
    })
  })
})


examples/configurations.js
var singleServer = {
  "db" : "sessions",
  "collection" : "express_sessions",
  "host" : "localhost",
  "port" : 27017
}

var replicaSet = {
  "collection" : "express_sessions",
  "stringify": false,
  "db": {
    "name" : "sessions",
    "servers" : [
      {
        "host" : "localhost",
        "port" : 27017,
        "options" : {
          "autoReconnect" : false,
          "poolSize" : 200,
          "socketOptions" : {
            "timeout" : 0,
            "noDelay" : true,
            "keepAlive" : 1,
            "encoding" : "utf8"
          }
        }
      },
      {
        "host" : "localhost",
        "port" : 27018,
        "options" : {
          "autoReconnect" : false,
          "poolSize" : 200,
          "socketOptions" : {
            "timeout" : 0,
            "noDelay" : true,
            "keepAlive" : 1,
            "encoding" : "utf8"
          }
        }
      },
      {
        "host" : "localhost",
        "port" : 27019,
        "options" : {
          "autoReconnect" : false,
          "poolSize" : 200,
          "socketOptions" : {
            "timeout" : 0,
            "noDelay" : true,
            "keepAlive" : 1,
            "encoding" : "utf8"
          }
        }
      }
    ]
  }
}

lib/connect-mongostore.js
'use strict'

var mongo = require('mongodb')
  , url = require('url')
  , util = require('util')
  , _ = require('lodash')


/**
 * Default options
 */
var defaultOptions = {
  'host': '127.0.0.1',
  'port': 27017,
  'stringify': false,
  'collection': 'sessions',
  'expireAfter': 1000 * 60 * 60 * 24 * 14, // 2 weeks
  'autoReconnect': false,
  'ssl': false,
  'w': 1
}



function getStore(Store) {
  /**
   * Initialize MongoStore with the given `options`.
   *
   * @param {Object} options
   */
  function MongoStore (options) {
    Store.call(this, this.options)

    if (typeof options == 'string') {
      this.options = parseUrl(options)
    } else {
      this.options = _.clone(options || {})
      this.options.expireAfter = this.options.expireAfter || defaultOptions.expireAfter
    }

    if (!this.options.hasOwnProperty('stringify')) {
      this.options.stringify = defaultOptions.stringify
    }

    this.collectionName = this.options.collection || defaultOptions.collection
    this.db = setupDb(this.options)
  }



  /**
   * Inherit from `Store`.
   */
  util.inherits(MongoStore, Store)



  MongoStore.prototype.serialize = function (obj) {
    if (this.options.stringify) {
      return JSON.stringify(obj)
    }
    return obj
  }


  MongoStore.prototype.deserialize = function (str) {
    if (this.options.stringify) {
      return JSON.parse(str)
    }
    return str
  }


  MongoStore.prototype.getDatabase = function (cb) {
    var self = this

    self.db.open(function (err, db) {
      if (err) throw new Error('Error connecting to database:\n' + err.message + '\n' + err.stack + '\n')

      if (self.options.username && self.options.password) {
        db.authenticate(self.options.username, self.options.password, function () {
          self.getCollection(cb)
        })
      } else {
        self.getCollection(cb)
      }
    })
  }



  /**
   * Returns a MongoDB collection.
   *
   * @param cb
   */
  MongoStore.prototype.getCollection = function (cb) {
    var self = this

    if (self.collection) {
      cb && cb(self.collection)
      return
    }

    if (!self.db.openCalled) return self.getDatabase(cb)

    self.db.collection(self.collectionName, function (err, collection) {
      if (err) throw new Error('Error getting collection: ' + self.collectionName + ' <' + err + '>')

      self.collection = collection

      self.collection.ensureIndex({'expires': 1}, {'expireAfterSeconds': 0}, function (err, result) {
        if (err) throw new Error('Error setting TTL index on collection : ' + self.collectionName + ' <' + err + '>')
        cb && cb(self.collection)
      })
    })
  }



  /**
   * Attempt to fetch session by the given `sid`.
   *
   * @param {String} sid
   * @param {Function} cb
   * @api public
   */
  MongoStore.prototype.get = function (sid, cb) {
    var self = this

    this.getCollection(function (collection) {
      collection.findOne({_id: sid}, function (err, session) {
        if (err) {
          cb && cb(err)
          return
        }

        if (!session) {
          cb && cb()
          return
        }

        if (!session.expires || new Date < session.expires) return cb(null, self.deserialize(session.session))
        self.destroy(sid, cb)
      })
    })
  }



  /**
   * Commit the given `session` object associated with the given `sid`.
   *
   * @param {String} sid
   * @param {Object} session
   * @param {Function} cb
   * @api public
   */
  MongoStore.prototype.set = function (sid, session, cb) {
    var s = {'_id': sid, 'session': this.serialize(session)}

    if (session && session.cookie && session.cookie._expires) {
      s.expires = new Date(session.cookie._expires)
    } else {
      s.expires = getFutureDate(this.options.expireAfter)
    }

    this.getCollection(function (collection) {
      collection.update({'_id': sid}, s, {'upsert': true, 'safe': true}, function (err, data) {
        cb(err, data)
      })
    })
  }



  /**
   * Destroy the session associated with the given `sid`.
   *
   * @param {String} sid
   * @param {Function} cb
   * @api public
   */
  MongoStore.prototype.destroy = function (sid, cb) {
    this.getCollection(function (collection) {
      collection.remove({_id: sid}, function () {
        cb && cb()
      })
    })
  }



  /**
   * Fetch number of sessions.
   *
   * @param {Function} cb
   * @api public
   */
  MongoStore.prototype.length = function (cb) {
    this.getCollection(function (collection) {
      collection.count({}, function (err, count) {
        cb && cb(err, count)
      })
    })
  }



  /**
   * Clear all sessions.
   *
   * @param {Function} cb
   * @api public
   */
  MongoStore.prototype.clear = function (cb) {
    this.getCollection(function (collection) {
      collection.drop(function () {
        cb && cb()
      })
    })
  }


  return MongoStore
}



/**
 * Returns a data in the future. By default,
 * returns now + 2 weeks.
 *
 * @param {Date} offset
 * @returns {Date}
 */
function getFutureDate(offset) {
  return new Date(Date.now() + offset)
}



/**
 * Parse a database URL.
 *
 * @param {String} path
 * @returns {Object}
 */
function parseUrl(path) {
  var parsed = {}

  var dbUri = url.parse(path)

  if (dbUri.port) {
    parsed.port = parseInt(dbUri.port)
  }

  if (dbUri.pathname != undefined) {
    var pathname = dbUri.pathname.split('/')

    if (pathname.length >= 2 && pathname[1]) {
      parsed.db = pathname[1]
    }

    if (pathname.length >= 3 && pathname[2]) {
      parsed.collection = pathname[2]
    }
  }

  if (dbUri.hostname != undefined) {
    parsed.host = dbUri.hostname
  }

  if (dbUri.auth != undefined) {
    var auth = dbUri.auth.split(':')

    if (auth.length >= 1) {
      parsed.username = auth[0]
    }

    if (auth.length >= 2) {
      parsed.password = auth[1]
    }
  }

  return parsed
}



/**
 * Instantiate database instance.
 *
 * @param opts
 * @returns {*}
 */
function setupDb(opts) {
  if (opts.mongooseConnection) return dbFromMongooseConnection(opts.mongooseConnection, opts)
  if (!opts.db) throw new Error('Required MongoStore option `db` missing')
  if (typeof opts.db == 'object' && opts.db.databaseName) return opts.db // Assume it's an instantiated DB Object

  if (Array.isArray(opts.db.servers)) {
    var serverArray = []

    opts.db.servers.forEach(function (server) {
      var serverOptions = server.options || {}
      var newServer = new mongo.Server(server.host, server.port, serverOptions)
      serverArray.push(newServer)
    })

    return new mongo.Db(opts.db.name, new mongo.ReplSetServers(serverArray), opts.db.replicaSetOptions || {'w': 1})
  }

  if (typeof opts.db != 'string') throw new Error('`db` option must be a string, array or a database instance.')

  return new mongo.Db(
    opts.db, new mongo.Server(
      opts.host || defaultOptions.host,
      opts.port || defaultOptions.port,
      {
        auto_reconnect: opts.autoReconnect || defaultOptions.autoReconnect,
        ssl: opts.ssl || defaultOptions.ssl
      }),
    { w: opts.w || defaultOptions.w })
}



/**
 * Make a DB instance from mongoose connection.
 * @param mongooseConnection
 * @param opts
 * @returns {Object} mongo-native Server instance
 */
function dbFromMongooseConnection(mongooseConnection, opts) {
  if (mongooseConnection.user && mongooseConnection.pass) {
    opts.username = mongooseConnection.user
    opts.password = mongooseConnection.pass
  }

  return new mongo.Db(
    mongooseConnection.db.databaseName,
    new mongo.Server(mongooseConnection.db.serverConfig.host,
      mongooseConnection.db.serverConfig.port,
      mongooseConnection.db.serverConfig.options),
    { w: opts.w || defaultOptions.w }
  )
}



module.exports = function (connect) {
  return getStore(connect.Store? connect.Store : connect.session.Store)
}
